File: Scripts/Room/CombatRoomHandler.cs
using Roguelike.Data;
using RoguelikeMapGen;
using System;
using System.Collections.Generic;
using System.Linq;

namespace Roguelike.Logic.Handlers
{
    public class CombatRoomHandler : IRoomHandler
    {
        public void Execute(GameRun run, Room room)
        {
            var enemyTemplates = GenerateEncounter(run, room.StarRating);

            if (!enemyTemplates.Any())
            {
                run.CurrentState = GameState.OnMap;
                return;
            }

            run.CurrentCombat = new CombatManager(run.TheHero, enemyTemplates, run.Rng, run.EffectPool.GetEffect);
            run.CurrentCombat.StartCombat();
            run.CurrentState = GameState.InCombat;
        }

        /// <summary>
        /// A method for generating enemy encounters based on a star rating.
        /// </summary>
        private List<EnemyData> GenerateEncounter(GameRun run, int starRating)
        {
            var encounter = new List<EnemyData>();

            var leader = run.EnemyPool.GetRandomEnemyOfStar(starRating, run.Rng);
            if (leader == null)
            {
                Console.WriteLine($"CRITICAL ERROR: No enemies found for Star Rating {starRating}. Aborting encounter generation.");
                return encounter;
            }
            encounter.Add(leader);

            int minionCount = 0;
            int maxMinionStar = 1;

            switch (starRating)
            {
                case 1: // Easy monster
                    minionCount = run.Rng.Next(0, 2); // 0 or 1 minion
                    maxMinionStar = 1;
                    break;
                case 2: // Harder monster
                    minionCount = run.Rng.Next(1, 3); // 1 or 2 minions
                    maxMinionStar = 1;
                    break;
                case 3: // Easy elite
                    minionCount = run.Rng.Next(1, 2); // 1 minion
                    maxMinionStar = 2;
                    break;
                case 4: // Harder elite
                    minionCount = run.Rng.Next(1, 3); // 1 or 2 minions
                    maxMinionStar = 3;
                    break;
            }

            for (int i = 0; i < minionCount; i++)
            {
                var minion = run.EnemyPool.GetRandomEnemyBelowStar(maxMinionStar + 1, run.Rng);
                if (minion != null)
                {
                    encounter.Add(minion);
                }
            }

            return encounter;
        }

        public static void GenerateVictoryRewards(GameRun run)
        {
            var room = run.TheMap.GetCurrentRoom();
            var hero = run.TheHero;
            int n = room.StarRating;

            if (hero.Relics.Any(r => r.Id == "blood_vial"))
            {
                hero.Heal(2);
            }

            // Gold Formula: e^(3 + k*n) + 10 with 0 < k < 1
            double goldCalc = Math.Exp(3 + (0.5 * n)) + 10;
            int goldReward = (int)Math.Floor(goldCalc);
            hero.CurrentGold += goldReward;

            run.CardRewardChoices.Clear();
            
            run.CardRewardChoices.Add(run.CardPool.GetRandomCardOfStar(n, run.Rng));

            int lowerStarLimit = (n == 1) ? 1 : n; 
            for (int i = 0; i < 2; i++)
            {
                int limit = (n == 1) ? 1 : n - 1;
                run.CardRewardChoices.Add(run.CardPool.GetRandomCardUpToStar(limit, run.Rng));
            }
            
            if (hero.Relics.Any(r => r.Id == "prayer_wheel"))
            {
                int limit = (n == 1) ? 1 : n - 1;
                run.CardRewardChoices.Add(run.CardPool.GetRandomCardUpToStar(limit, run.Rng));
            }
            
            run.CardRewardChoices.RemoveAll(c => c == null);

            run.RelicRewardChoice = run.RelicPool.GetRandomRelicOfStar(n, run.Rng);
            
            if (room.Type == RoomType.Elite && hero.Relics.Any(r => r.Id == "black_star"))
            {
                var extraRelic = run.RelicPool.GetRandomRelicOfStar(n, run.Rng);
                if (extraRelic != null)
                {
                    hero.Relics.Add(extraRelic);
                }
            }

            run.CurrentState = GameState.AwaitingReward;
            run.CurrentCombat = null;
        }
    }
}

File: python_roguelike/core/rooms/combat_room_handler.py
import math
from typing import List
from .i_room_handler import IRoomHandler
from ..combat_manager import CombatManager
from ...data.combatant.enemy_data import EnemyData
from ...data.enums import GameState


class CombatRoomHandler(IRoomHandler):
    def execute(self, run, room):
        enemy_templates = self._generate_encounter(run, room.star_rating)
        if not enemy_templates:
            run.current_state = GameState.OnMap
            return

        run.current_combat = CombatManager(
            run.the_hero,
            enemy_templates,
            run.rng,
            run.effect_pool.get_effect
        )
        run.current_combat.start_combat()
        run.current_state = GameState.InCombat

    def _generate_encounter(self, run, star_rating: int) -> List[EnemyData]:
        encounter = []

        leader = run.enemy_pool.get_random_enemy_of_star(star_rating, run.rng)
        if leader is None:
            print(f"CRITICAL ERROR: No enemies found for Star Rating {star_rating}.")
            return encounter
        encounter.append(leader)

        minion_count = 0
        max_minion_star = 1

        if star_rating == 1:
            minion_count = 0
            max_minion_star = 1
        elif star_rating == 2:
            minion_count = run.rng.randint(1, 2)
            max_minion_star = 1
        elif star_rating == 3:
            minion_count = run.rng.randint(1, 1)
            max_minion_star = 2
        elif star_rating == 4:
            minion_count = run.rng.randint(1, 2)
            max_minion_star = 3

        for _ in range(minion_count):
            minion = run.enemy_pool.get_random_enemy_below_star(max_minion_star + 1, run.rng)
            if minion is not None:
                encounter.append(minion)

        return encounter

    @staticmethod
    def generate_victory_rewards(run):
        room = run.the_map.get_current_room()
        hero = run.the_hero
        n = room.star_rating

        gold_calc = math.exp(3 + 0.5 * n) + 10
        gold_reward = int(math.floor(gold_calc))
        hero.current_gold += gold_reward

        run.card_reward_choices.clear()

        top_card = run.card_pool.get_random_card_of_star(n, run.rng)
        if top_card:
            run.card_reward_choices.append(top_card)

        for _ in range(2):
            limit = 1 if n == 1 else n - 1
            extra = run.card_pool.get_random_card_up_to_star(limit, run.rng)
            if extra:
                run.card_reward_choices.append(extra)

        run.card_reward_choices = [c for c in run.card_reward_choices if c is not None]
        run.relic_reward_choice = run.relic_pool.get_random_relic_of_star(n, run.rng)

        run.current_state = GameState.AwaitingReward
        run.current_combat = None
